### PR TITLE
Auth java code

### PR DESCRIPTION
**Pull request by MarcRoopchand**
AuthService
- Permission check completed 
- Also Added Caching

### CODE REVIEW

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    Group group = (Group) user.getGroup();
+    if (group == null) {
+      return false;
+    }
+
+    /**
+     * Check if data is already cached, if it is return this data
+     * If not, save it to the maps
+     * */
+    if (permissionMap.containsKey(permission.getName())) {
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: See the https://github.com/foam-framework/foam/wiki/FOAM2-Style-Guidelines 

This line should be:
if ( permissionMap.containsKey(permission.getName()) ) {

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      ConcurrentHashMap<String, Boolean> userMap = permissionMap.get(permission.getName());
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: If you just declare the type to be Map<String, Boolean> then if/when you change the type in the future you wouldn't have the change the type here. Code to the interface not the implementation.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+      if (userMap.containsKey(user.getId())) {
+        return userMap.get(user.getId());
+      }
+
+      Boolean permissionCheck = group.implies(permission.getName());
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: Should be 'boolean', not 'Boolean'.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      userMap.put(user.getId(), permissionCheck);
+      permissionMap.put(permission.getName(), userMap);
+      return permissionCheck;
+    }
+    else {
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: No need for an 'else' if you've already returned. 
if ( ) {

  return;
}  else {
  foo();
}

Can just be written:

if ( ) {

  return;
}

foo();

Which reduces nesting.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      ConcurrentHashMap<String, Boolean> userMap = new ConcurrentHashMap<>();
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: If you're not using any specific parts of the ConcurrentHashMap class, just declare the type as Map.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+      return permissionCheck;
+    }
+  }
+
+  /**
+   * On logout, walk through entire map and remove all instances of the user
+   * */
+  @Override
+  public X logout(X x) {
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: I think the return type should be 'void'.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    if (x == null) {
+      return null;
+    }
+
+    User user = (User) x.get("user");
+    if (user == null) {
+      return null;
+    }
+
+    for (String key: permissionMap.keySet()) {
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: This seems very inefficient. If you had a map of maps then you could just remove one entry without iterating.

File: src/foam/nanos/auth/Group.js [hunk 8]
```diff
@@ -38,5 +44,25 @@ foam.CLASS({
...
+      args: [
+        {
+          name: 'name',
+          javaType: 'String'
+        }
+      ],
+      javaCode:
+        'if (getPermissions() == null) {return false;} \n' +
+        'for (int i = 0; i < permissions_.length; i++) { \n' +
+        '  if (permissions_[i].getId().equals(name)) { \n' +
```
Comment (kgrgreer) [inline src/foam/nanos/auth/Group.js]: Don't use equality, create a java.security.Permission and call its implies() method. This adds wildcard support.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java (hunk 7 above)
Comment (MarcRoopchand) [reply java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: @kgrgreer What if your user belongs to multiple permissions? 
Say your cached data looked something like this 
{
  permission1: {
    user1: true
  }
  permission2: {
    user1: true
  }
}

Now a user logs out, and I need to remove this user from all the permissions.

File: java_src/foam/nanos/auth/LRULinkedHashMap.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package foam.nanos.auth;
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/LRULinkedHashMap.java]: Maybe put in foam.nanos.util, since it will be useful for others also.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+   *  }
+   *  ...
+   *}
+   */
+  protected Map<String, Map<String, Boolean>> permissionMap;
+
+  @Override
+  public void start() {
+    super.start();
+    permissionMap = new LRULinkedHashMap<>(1000);
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: This could just be set it permissionMap's declaration.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+    Group group = (Group) user.getGroup();
+    if ( group == null ) return false;
+
+    /**
+     * Check if data is already cached, if it is return this data
+     * If not, save it to the maps
+     * */
+    if ( permissionMap.containsKey(permission.getName()) ) {
+      Map<String, Boolean> userMap = permissionMap.get(permission.getName());
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: The code below could be restructured to avoid duplication.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+   * On logout, walk through entire map and remove all instances of the user
+   * */
+  @Override
+  public void logout(X x) {
+    if ( x == null ) return;
+
+    User user = (User) x.get("user");
+    if ( user == null ) return;
+
+    for ( String key: permissionMap.keySet() ) {
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: Won't this be a lot more efficient if you make the userMap be the top-level collection instead?

File: java_src/foam/nanos/auth/UserAndGroupAuthService.java [hunk 13]
```diff
@@ -195,7 +171,7 @@ public Boolean validateUser(User user) throws IllegalStateException {
    * Just return a null user for now. Not sure how to handle the cleanup
    * of the current context
    */
-  public X logout(X x) {
-    return this.getX().put("user", null);
+  public void logout(X x) {
+    this.getX().put("user", null);
```
Comment (kgrgreer) [inline java_src/foam/nanos/auth/UserAndGroupAuthService.java]: Contexts are immutable, so this line doesn't do anything. Which is good, cause it isn't needed.

File: src/foam/nanos/auth/Group.js [hunk 14]
```diff
@@ -38,5 +44,25 @@ foam.CLASS({
...
+      args: [
+        {
+          name: 'permission',
+          javaType: 'java.security.Permission'
+        }
+      ],
+      javaCode:
+        'if ( getPermissions() == null ) return false; \n' +
+        'for ( int i = 0; i < permissions_.length; i++ ) { \n' +
+        '  if ( permission.implies(new javax.security.auth.AuthPermission(permissions_[i].getId())) ) { \n' +
```
Comment (kgrgreer) [inline src/foam/nanos/auth/Group.js]: I think you have the arguments of implies the wrong way round.

File: java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java (hunk 12 above)
Comment (MarcRoopchand) [reply java_src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: I can do that but permission map contains multiple userMaps. 
Don't I still need to find which userMap I should be accessing?

File: src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+   *    ...
+   *  }
+   *  ...
+   *}
+   */
+  protected Map<String, Map<String, Boolean>> userMap = new LRULinkedHashMap<>(1000000);
+  protected Map<String, Boolean> permissionMap;
+
+  @Override
+  public void start() {
```
Comment (kgrgreer) [inline src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: Why is this needed?

File: src/foam/nanos/auth/CachedUserAndGroupAuthService.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    if ( user == null ) return false;
+
+    Group group = (Group) user.getGroup();
+    if ( group == null ) return false;
+
+    if ( userMap.containsKey(user.getId()) ) {
+      permissionMap = userMap.get(user.getId());
+    }
+    else {
+      permissionMap = new LRULinkedHashMap<>(1000);
```
Comment (kgrgreer) [inline src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: I think 1000 needs to be larger, or even just make this a regular HashMap. A user can easily be checked against thousands of permissions if you have one for each menu and one for each property and mode. The users will get cleaned up anyway, so this is redundant.

File: src/foam/nanos/auth/Group.js [hunk 17]
```diff
@@ -38,5 +44,25 @@ foam.CLASS({
...
+    {
+      name: 'implies',
+      javaReturns: 'Boolean',
+      args: [
+        {
+          name: 'permission',
+          javaType: 'java.security.Permission'
+        }
+      ],
+      javaCode:
```
Comment (kgrgreer) [inline src/foam/nanos/auth/Group.js:58]: Use multi-line backquote (`) to avoid having to concatenate lines together.

File: src/foam/nanos/auth/CachedUserAndGroupAuthService.java (hunk 15 above)
Comment (MarcRoopchand) [reply src/foam/nanos/auth/CachedUserAndGroupAuthService.java]: I assumed that we will need to start this service from Boot.

How do you envision this caching fitting into the system ?

File: src/foam/nanos/auth/UserAndGroupAuthServiceTest.java [hunk 18]
```diff
@@ -1,72 +1,164 @@
...
+  public void addTestUsers() {
+    System.out.println("Registering " + numUsers + " Users");
+    long startTime  = System.nanoTime();
+    ListSink sink   = (ListSink) groupDAO_.select(new ListSink());
+
+    /**
+     * For each user, randomly select a group from the groups created
+     * and assign the user to this group
+     * */
+    for ( int i = 0; i < numUsers; i++ ) {
```
Comment (kgrgreer) [inline src/foam/nanos/auth/UserAndGroupAuthServiceTest.java]: See FOAM style-guide for spacing:
for ( int i = 0 ; i < numUsers ; i++ ) {

File: src/foam/util/LRULinkedHashMap.java [hunk 19]
```diff
@@ -0,0 +1,20 @@
+package foam.nanos.util;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * Created by marcroopchand on 2017-05-24.
+ */
+public class LRULinkedHashMap<K, V> extends LinkedHashMap<K, V> {
+  private final int MAX_SIZE;
```
Comment (kgrgreer) [inline src/foam/util/LRULinkedHashMap.java]: Rename MAX_SIZE to maxSize_. CONSTANT naming convention implies it's a constant, but it's just final.

File: src/foam/nanos/auth/Group.js (hunk 17 above)
Comment (MarcRoopchand) [reply src/foam/nanos/auth/Group.js:58]: We should add this to the style guide as well. Seems throughout the project all javaCode has concatenation
